### ISSUE TITLE
[Question] Supporting Multiple OpenAPI Specs

### ISSUE THREAD

**Issue by brandonmaul** [apple/swift-openapi-generator#177]
Hello, I'm curious if this project supports the use of having multiple OpenAPI specs in a project, and being able to define which one to use at build time. 

This is because I have a project which currently has 3 maturity levels of the API (dev, stage, prod), and to configure a build of the app for the right environment, we have different build configurations in Xcode.

I would ideally like to be able to specify which OpenAPI document to point to at build time / at the time of which the plugin runs, based on the build configuration. I would have all 3 specs in the repo, one for each maturity.

**Comment by czechboy0** [comment]
Hi @brandonmaul,

sure, you can use many OpenAPI documents in the same _package_ or _project_, but note that you cannot generate code from multiple OpenAPI documents _into_ one _target_. You'd want to create one target per OpenAPI doc, and then your other packages can import the ones with the generated code.

In your case, that could mean, for example, setting up a local package that looks like:

```
Package/Sources
  - DevAPI
    - openapi.yaml
    - openapi-generator-config.yaml
  - StageAPI
    - openapi.yaml
    - openapi-generator-config.yaml
  - ProdAPI
    - openapi.yaml
    - openapi-generator-config.yaml
```

And then e.g. your app target can just `import DevAPI` or `import StageAPI`, etc based on what it needs.

So all 3 can coexist as separate targets, that might be easier to maintain than somehow trying to dynamically replace the OpenAPI document dynamically at compile time.

Hope this helps!

**Comment by brandonmaul** [comment]
Thank you Honza!

Unfortunately for me, my team has requirements that don't allow us to break up the deliverable (the App in this case) into multiple targets.

As I've sat here thinking about how to solve it, I believe I could add a build phase to overwrite the OpenAPI spec document before the swift-openapi-generator plugin runs, effective replacing the text of the spec based on the build configuration.

This is quite a "hacky" workaround though. And since I've not done it yet, im not even sure it will work.

**Comment by czechboy0** [comment]
You could use a symlink, which is a common practice for folks who want to keep their OpenAPI document at the top of their repo, but have their Swift package target consume it as well. In your case, your script could update the symlink to point to the right file.

Now, while the requirement to not have code in more than one target is unfortunate, I don't think we can help work around it except for the symlink suggestion, and possibly other workarounds. Making it possible to generate multiple OpenAPI documents into a single target is explicitly out of scope for this project, as the Swift ecosystem already covers the use case of separating groups of files into shared scopes, and those are targets. To work around this, we'd essentially have to create a targets-like feature in Swift OpenAPI Generator, which would be a lot of work.

Closing this issue as I think we covered the possibilities of what Swift OpenAPI Generator can help you with, good luck, hopefully you can get it to work!

**Comment by brandonmaul** [comment]
Update:

Thanks for the tip on symlinks. I found that was a solution to this. I have 3 OpenAPI files in my project, and a symlink that gets updated at pre-build time.
<img width="193" alt="Screenshot 2023-08-08 at 10 33 55 AM" src="https://github.com/apple/swift-openapi-generator/assets/22066161/b834befb-b765-4d40-9139-3514f2d29a33">

Since each other build configurations have a corresponding Xcode schema, I could simply put the symlink update script as a build pre-action for each schema. For example, here's the "demo" schema's build pre-action. 

```
readlink -f "$SRCROOT/AndiumExplorer/openapi.json"

ln -sf "$SRCROOT/AndiumExplorer/openapi-demo.json" "$SRCROOT/AndiumExplorer/openapi.json"

readlink -f "$SRCROOT/AndiumExplorer/openapi.json"
```

I also found that the symlink needs to exist before you build entirely, because Xcode runs your SPM dependencies pre-build phases before running your scheme's "Pre-Actions". Without the symlink existing in the first place, the `swift-openapi-generator` would throw an error that no openapi.json file existed before my script was able to run.

Anyway, thanks @czechboy0 for the quick communication. I know I have a weird use case, but hopefully this thread could be useful in the event someone else runs into this situation.
